components/file_system/src/rate_limiter.rs

```diff
@@ -505,39 +353,120 @@ impl IORateLimiter {
+                    );
+                    warn!("unprocessed read bytes: {:?}", (t, bytes));
+                    request_imp!(self, p, bytes, async);
+                }
+                ctx.outstanding_read_bytes += bytes;
+                if ctx.outstanding_read_bytes >= batch {
+                    let true_bytes = io_stats::fetch_thread_io_bytes().read;
+                    warn!(
+                        "true bytes: {:?}",
+                        (
+                            ctx.io_type,
+                            true_bytes - ctx.total_read_bytes,
+                            ctx.outstanding_read_bytes
+                        )
+                    );
+                    request_imp!(
+                        self,
+                        priority,
+                        std::cmp::min(
+                            true_bytes - ctx.total_read_bytes,
+                            ctx.outstanding_read_bytes,
+                        ),
+                        async
+                    );
+                    ctx.outstanding_read_bytes = 0;
+                    ctx.total_read_bytes = true_bytes;
+                }
+            }
         }
+        io_stats::set_io_context(ctx);
         if let Some(stats) = &self.stats {
-            stats.record(io_type, io_op, bytes);
+            stats.record(ctx.io_type, io_op, bytes);
         }
         bytes
     }
 
     #[cfg(test)]
-    fn request_with_skewed_clock(&self, io_type: IOType, io_op: IOOp, mut bytes: usize) -> usize {
+    fn request_for_test(&self, io_type: IOType, io_op: IOOp, bytes: usize) -> usize {
         if self.mode.contains(io_op) {
-            bytes = self.throughput_limiter.request_with_skewed_clock(
-                IOPriority::unsafe_from_u32(
-                    self.priority_map[io_type as usize].load(Ordering::Relaxed),
-                ),
-                bytes,
+            let priority = IOPriority::unsafe_from_u32(
+                self.priority_map[io_type as usize].load(Ordering::Relaxed),
             );
+            request_imp!(self, priority, bytes, sync);
         }
         if let Some(stats) = &self.stats {
             stats.record(io_type, io_op, bytes);
         }
         bytes
     }
+
+    /// Updates and refills I/O budgets for next epoch based on I/O priority.
+    /// This method is a no-op when high-priority I/O rate limit equals zero.
+    /// Here we provide best-effort priority control:
+    /// 1) Limited I/O budget is assigned to lower priority to ensure higher priority can at least
+    ///    consume the same I/O amount as the last few epochs without breaching global threshold.
+    /// 2) Higher priority may temporarily use lower priority's I/O budgets. When this happens,
+    ///    total I/O flow could exceed global threshold.
+    /// 3) Highest priority I/O alone must not exceed global threshold (in strict mode).
+    fn refill(&self, locked: &mut IORateLimiterInner, now: Instant) {
```

Comment:
please do not move the place of it, I can hardly know what has changed.

Reply:
It's hard to do that. The old `PriorityBasedIORateLimiter` is completed removed.

I created a temporary commit to show the diff here: https://github.com/tabokie/tikv/commit/1ba21fdf40175d24cd23e3b16e3959680ae86550